src/Hint/Lambda.hs

```diff
@@ -86,96 +93,179 @@ yes = map (\f -> dataDir </> f) dataFiles -- (dataDir </>)
-              munge ident p = PVar (ann p) (Ident (ann p) [ident])
-              subts fps b = ("body", toSS b) : zipWith (\x y -> ([x],y)) ['a'..'z'] (map toSS fps)
-              s1 = subts finalpats body
-              --s2 = subts pats2 bod2
-              t1 = template finalpats body
-              --t2 = template pats2 bod2
-
-lambdaDecl _ = []
-
-setSpanInfoEnd ssi (line, col) = ssi{srcInfoSpan = (srcInfoSpan ssi){srcSpanEndLine=line, srcSpanEndColumn=col}}
-
-
-etaReduce :: [Pat_] -> Exp_ -> ([Pat_], Exp_)
-etaReduce ps (App _ x (Var _ (UnQual _ (Ident _ y))))
-    | ps /= [], PVar _ (Ident _ p) <- last ps, p == y, p /= "mr", y `notElem` vars x
-    , not $ any isQuasiQuote $ universe x
-    = etaReduce (init ps) x
-etaReduce ps (InfixApp a x (isDol -> True) y) = etaReduce ps (App a x y)
-etaReduce ps x = (ps,x)
+lambdaHint :: DeclHint'
+lambdaHint _ _ x
+    =  concatMap (uncurry lambdaExp') (universeParentBi x)
+    ++ concatMap lambdaDecl' (universe x)
+
+-- TODO: handle PatBinds
+lambdaDecl' :: LHsDecl GhcPs -> [Idea]
+lambdaDecl'
+    o@(LL loc1 (ValD _
+        origBind@(FunBind {fun_matches =
+            (MG {mg_alts =
+                LL _ [LL _ (Match _ ctxt pats (GRHSs _ [LL loc2 (GRHS _ [] origBody)] bind))]})})))
+    | LL _ (EmptyLocalBinds noExt) <- bind
+    , isLambda $ fromParen' origBody
+    , null (universeBi pats :: [HsExpr GhcPs])
+    = [warn' "Redundant lambda" o (gen pats origBody) [Replace Decl (toSS' o) s1 t1]]
+    | length pats2 < length pats, pvars' (drop (length pats2) pats) `disjoint` varss' bind
+    = [warn' "Eta reduce" (reform' pats origBody) (reform' pats2 bod2)
+          [ -- Disabled, see apply-refact #3
+            -- Replace Decl (toSS' $ reform' pats origBody) s2 t2]]
+          ]]
+    where reform' :: [LPat GhcPs] -> LHsExpr GhcPs -> LHsDecl GhcPs
+          reform' ps b = LL loc $ ValD noExt $
+            origBind
+              {fun_matches = MG noExt (noLoc [noLoc $ Match noExt ctxt ps $ GRHSs noExt [noLoc $ GRHS noExt [] b] $ noLoc $ EmptyLocalBinds noExt]) Generated}
+
+          loc = combineSrcSpans loc1 loc2
+
+          gen :: [Pat GhcPs] -> LHsExpr GhcPs -> LHsDecl GhcPs
+          gen ps = uncurry reform' . fromLambda' . lambda ps
 
+          (finalpats, body) = fromLambda' . lambda pats $ origBody
+          (pats2, bod2) = etaReduce' pats origBody
+          template fps = unsafePrettyPrint $ reform' (zipWith munge' ['a'..'z'] fps) varBody
+          subts fps b = ("body", toSS' b) : zipWith (\x y -> ([x],y)) ['a'..'z'] (map toSS' fps)
+          s1 = subts finalpats body
+          --s2 = subts pats2 bod2
+          t1 = template finalpats
+          --t2 = template pats2 bod2
+lambdaDecl' _ = []
+
+
+etaReduce' :: [Pat GhcPs] -> LHsExpr GhcPs -> ([Pat GhcPs], LHsExpr GhcPs)
+etaReduce' (unsnoc -> Just (ps, view' -> PVar_' p)) (LL _ (HsApp _ x (view' -> Var_' y)))
+    | p == y
+    , p /= "mr"
+    -- TODO: what is this "mr" thing??
```

Comment:
Why do we not eta-reduce variables named `mr`? :thinking:

Reply:
Because `mr` stands for monomorphism restriction and some people type `foo mr = bar mr` to indicate that this variable can't be eta reduced or you trip into that restriction. From the days before easy ignoring, and I suspect no one knows about it - feel free to delete.

Reply:
Deleted it.